File: backend/data/skill_store/imported/nanoresearch/bio-workflows-scrnaseq-pipeline/scripts/python/_skill_registry.py
"""Skill Registry — Unified dependency resolution for CellClaw pipeline skills.

Resolution order (deterministic, transparent):
    1. Environment variable: CELLCLAW_SKILL_<SKILL_NAME>
    2. Registry file: ~/.cellclaw/skills.json
    3. Fallback relative path from this script's parent

Usage:
    from _skill_registry import resolve_skill_path, import_skill_module
    data_io_path = resolve_skill_path("bio-single-cell-data-io")
    adata = import_skill_module("bio-single-cell-data-io", "samplesheet").load_from_samplesheet(path)
"""

import json
import os
import sys
from pathlib import Path


def _get_env_var(skill_name: str) -> str:
    """Convert skill name to env var key, e.g. bio-single-cell-data-io → CELLCLAW_SKILL_BIO_SINGLE_CELL_DATA_IO."""
    return f"CELLCLAW_SKILL_{skill_name.upper().replace('-', '_')}"


def _get_registry_file() -> Path:
    """Path to the central skills registry JSON."""
    env_registry = os.environ.get("CELLCLAW_SKILLS_REGISTRY")
    if env_registry:
        return Path(env_registry)
    return Path.home() / ".cellclaw" / "skills.json"


def _get_fallback_path(skill_name: str, subpath: str) -> Path:
    """Compute fallback relative path from this script's parent."""
    script_dir = Path(__file__).parent.resolve()
    return script_dir / f"../../../{skill_name}/{subpath}"


def _read_registry(skill_name: str, subpath: str) -> Path | None:
    """Try to resolve skill path from the central registry file."""
    registry_file = _get_registry_file()
    if not registry_file.exists():
        return None
    try:
        with open(registry_file, "r", encoding="utf-8") as f:
            registry = json.load(f)
    except (json.JSONDecodeError, OSError):
        return None

    entry = registry.get(skill_name)
    if not entry:
        return None

    # Entry can be a string (path) or a dict with 'path' key
    if isinstance(entry, dict):
        path_str = entry.get("path")
    else:
        path_str = str(entry)

    if not path_str:
        return None

    resolved = Path(path_str).expanduser().resolve() / subpath
    if resolved.exists():
        return resolved
    return None


def resolve_skill_path(skill_name: str, subpath: str = "scripts/python") -> Path:
    """Resolve skill path via env var → registry file → relative fallback.

    Parameters
    ----------
    skill_name : str
        Skill name, e.g. "bio-single-cell-data-io"
    subpath : str
        Subdirectory to append (e.g. "scripts/python", "scripts/r")

    Returns
    -------
    Path
        Absolute path to the skill's scripts directory.

    Raises
    ------
    FileNotFoundError
        If none of the three resolution methods succeed.
    """
    attempted = []

    # 1. Environment variable
    env_key = _get_env_var(skill_name)
    env_path = os.environ.get(env_key)
    if env_path:
        resolved = Path(env_path).expanduser().resolve() / subpath
        if resolved.exists():
            return resolved
        attempted.append(f"Env var {env_key}={env_path} (not found)")
    else:
        attempted.append(f"Env var {env_key} (not set)")

    # 2. Registry file
    registry_path = _read_registry(skill_name, subpath)
    if registry_path is not None:
        return registry_path
    attempted.append(f"Registry file {_get_registry_file()} (missing or no entry)")

    # 3. Fallback relative path
    fallback = _get_fallback_path(skill_name, subpath)
    if fallback.exists():
        return fallback.resolve()
    attempted.append(f"Fallback {fallback} (not found)")

    raise FileNotFoundError(
        f"Cannot resolve skill '{skill_name}' (subpath='{subpath}'). Tried:\n"
        + "\n".join(f"  {i + 1}. {a}" for i, a in enumerate(attempted))
        + "\n\nFix one of:\n"
        f"  • export {env_key}=/path/to/{skill_name}\n"
        f"  • Add to ~/.cellclaw/skills.json: {{\"{skill_name}\": \"/path/to/{skill_name}\"}}\n"
        f"  • Ensure {skill_name} is checked out at ../../../{skill_name}/"
    )


def import_skill_module(skill_name: str, module_name: str, subpath: str = "scripts/python"):
    """Dynamically import a module from a skill's scripts directory.

    Parameters
    ----------
    skill_name : str
        Skill name
    module_name : str
        Python module name (e.g. "samplesheet")
    subpath : str
        Subdirectory within the skill

    Returns
    -------
    module
        Imported module object

    Raises
    ------
    ImportError
        If the module cannot be imported
    """
    skill_path = resolve_skill_path(skill_name, subpath=subpath)
    path_str = str(skill_path)
    if path_str not in sys.path:
        sys.path.insert(0, path_str)

    try:
        return __import__(module_name)
    except ImportError as e:
        raise ImportError(
            f"Failed to import '{module_name}' from skill '{skill_name}' at {skill_path}: {e}"
        ) from e

File: data/skill_store/imported/nanoresearch/bio-workflows-scrnaseq-pipeline/scripts/r/_skill_registry.R
#' Skill Registry — Unified dependency resolution for CellClaw pipeline skills
#'
#' Resolution order (deterministic, transparent):
#'   1. Environment variable: CELLCLAW_SKILL_<SKILL_NAME>
#'   2. Registry file: ~/.cellclaw/skills.json
#'   3. Fallback relative path from this script's parent
#'
#' Usage:
#'   source("scripts/r/_skill_registry.R")
#'   data_io_dir <- resolve_skill_path("bio-single-cell-data-io", "scripts/r")
#'   source(file.path(data_io_dir, "samplesheet.R"))


# ---------------------------------------------------------------------------
# Helpers
# ---------------------------------------------------------------------------

.resolve_env_var_name <- function(skill_name) {
  #' Convert skill name to env var key.
  #' e.g. "bio-single-cell-data-io" -> "CELLCLAW_SKILL_BIO_SINGLE_CELL_DATA_IO"
  toupper(gsub("-", "_", skill_name))
}

.resolve_registry_file <- function() {
  #' Path to the central skills registry JSON.
  env_registry <- Sys.getenv("CELLCLAW_SKILLS_REGISTRY", unset = NA)
  if (!is.na(env_registry)) {
    return(env_registry)
  }
  file.path(path.expand("~"), ".cellclaw", "skills.json")
}

.resolve_from_registry <- function(skill_name, subpath) {
  #' Try to resolve skill path from the central registry file.
  registry_file <- .resolve_registry_file()
  if (!file.exists(registry_file)) {
    return(NULL)
  }

  registry <- tryCatch(
    jsonlite::fromJSON(registry_file),
    error = function(e) NULL
  )
  if (is.null(registry)) {
    return(NULL)
  }

  entry <- registry[[skill_name]]
  if (is.null(entry)) {
    return(NULL)
  }

  # Entry can be a string (path) or a list with 'path' key
  if (is.list(entry) && !is.null(entry$path)) {
    path_str <- entry$path
  } else {
    path_str <- as.character(entry)
  }

  resolved <- file.path(path_str, subpath)
  if (dir.exists(resolved)) {
    return(resolved)
  }
  return(NULL)
}

.resolve_fallback <- function(skill_name, subpath) {
  #' Compute fallback relative path from this script's parent.
  this_file <- sys.frame(1)$ofile
  if (is.null(this_file) || this_file == ".") {
    this_dir <- getwd()
  } else {
    this_dir <- dirname(this_file)
  }
  file.path(this_dir, "..", "..", "..", skill_name, subpath)
}


# ---------------------------------------------------------------------------
# Main resolver
# ---------------------------------------------------------------------------

resolve_skill_path <- function(skill_name, subpath = "scripts/r") {
  #' Resolve skill path via env var -> registry file -> relative fallback.
  #'
  #' @param skill_name Skill name, e.g. "bio-single-cell-data-io"
  #' @param subpath Subdirectory to append, e.g. "scripts/r"
  #' @return Absolute path to the skill's scripts directory
  #' @throws Error if none of the three methods succeed

  attempted <- c()

  # 1. Environment variable
  env_key <- paste0("CELLCLAW_SKILL_", .resolve_env_var_name(skill_name))
  env_path <- Sys.getenv(env_key, unset = NA)
  if (!is.na(env_path)) {
    resolved <- file.path(env_path, subpath)
    if (dir.exists(resolved)) {
      return(normalizePath(resolved))
    }
    attempted <- c(attempted, sprintf("Env var %s=%s (directory not found)", env_key, env_path))
  } else {
    attempted <- c(attempted, sprintf("Env var %s (not set)", env_key))
  }

  # 2. Registry file
  registry_path <- .resolve_from_registry(skill_name, subpath)
  if (!is.null(registry_path)) {
    return(normalizePath(registry_path))
  }
  attempted <- c(attempted, sprintf("Registry file %s (missing or no entry)", .resolve_registry_file()))

  # 3. Fallback relative path
  fallback <- .resolve_fallback(skill_name, subpath)
  if (dir.exists(fallback)) {
    return(normalizePath(fallback))
  }
  attempted <- c(attempted, sprintf("Fallback %s (not found)", fallback))

  # None succeeded — raise informative error
  msg <- paste0(
    "Cannot resolve skill '", skill_name, "' (subpath='", subpath, "'). Tried:\n",
    paste(sprintf("  %d. %s", seq_along(attempted), attempted), collapse = "\n"),
    "\n\nFix one of:\n",
    sprintf("  • Sys.setenv(%s = \"/path/to/%s\")\n", env_key, skill_name),
    sprintf("  • Add to ~/.cellclaw/skills.json: {\"%s\": \"/path/to/%s\"}\n", skill_name, skill_name),
    sprintf("  • Ensure %s is checked out at ../../../%s/", skill_name, skill_name)
  )
  stop(msg)
}
